### PR TITLE
User Request Improvements

### PR DESCRIPTION
**Pull request by tiferrei**
Clean-up of #19

### CODE REVIEW

**Comment by tiferrei** [comment]
@pietbrauer, we have a small problem, on both 3.0 and 2.3 branches, the podspec is still pointing to RequestKit 0.3.0, so it builds perfectly as source code, but when using TanukiKit in another project as a cocoa pod, it kinda breaks the Routers as the syntax changed a bit with the new RequestKit so it fails to compile. And to get things even worse, we can't specific branches for dependencies in the podspec, it only accepts releases.

For example, errors like:
Use of undeclared type 'RequestKitURLSession' and Type 'PublicKeyRouter' does not conform to protocol 'Router' as this version still required the params to be [String : String]

**Comment by pietbrauer** [comment]
I read it for the third time now, thanks! This is a problem on your app and specific to CocoaPods.

Add this to your Profile:

`pod "NBNRequestKit", git: "https://github.com/nerdishbynature/RequestKit.git", branch: "swift-2.3"`

Im sorry for being impolite but I don't have time at the moment to mentor you half of my day and it's stressing me out a bit.

**Comment by tiferrei** [comment]
That's fine, I completely understand it, it's just that it completely breaks the framework, making it useless and as you gave absolutely no sign of reading it, no monochromatic emoji, not even "got it" I though of keeping warning you about something that, as I said, breaks the framework, also, I had already tried that and it doesn't work because you can install whichever version of RequestKit you want, TanukiKit will still require 0.3.0, and only 0.3.0, as configured to by its podspec.
And me repeating the message was not meant to make sure you read it, if you notice, it is in each issue and pull request other contributors may come across if they have the same problem.

**Comment by pietbrauer** [comment]
Ok, here are a few things about open source:
1. People work for free on it
2. People mostly work in their free time
3. Everything happens asynchronous

You cannot request me to answer your personal project questions within 2 hours. That is not possible for anyone. If I have not answered in a week, it is ok for you to ping me, don't be pushy. Try to resolve the issue yourself, this is not specific to `TanukiKit`. There is no way to specify a branch in a podspec. The way I described, is the way to go.

If you still have problems: ask on stack overflow, thats what it is for and many more people will read it, if they run into the same issue.

I will look into your PullRequest at the end of the week.

**Comment by tiferrei** [comment]
I am not requesting anything, it's just good to keep record of how stuff works as it might be useful for other users. I completely understand that about open-source and that you work in this on your free-time and that you need to focus on your "making a living" work, I'm sorry if I didn't make that clear, I appreciate every piece of of this software and your work, and thank you for putting it open-source, I believe most people would just use it for them and not other people, so thank you very much, this is very kind of you to do, but I just wanted to do my part in helping you and not being another pain in the back, even if it is by just submitting issues or making small pull requests.

Thank you,
Tiago

File: TanukiKit/User.swift [hunk 1]
```diff
@@ -3,20 +3,60 @@ import RequestKit
...
+    public var colorSchemeId: Int?
+    public var projectsLimit: Int?
+    public var currentSignInAt: String?
+    public var canCreateGroup: Bool?
+    public var canCreateProject: Bool?
+    public var twoFactorEnabled: Bool?
+    public var external: Bool?
     public var privateToken: String?
-
+
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please remove the tab

File: TanukiKit/User.swift [hunk 2]
```diff
@@ -51,28 +90,28 @@ public extension TanukiKit {
 
 enum UserRouter: Router {
     case ReadAuthenticatedUser(Configuration)
-
+
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please remove the tab

File: TanukiKit/User.swift [hunk 3]
```diff
@@ -51,28 +90,28 @@ public extension TanukiKit {
...
     case ReadAuthenticatedUser(Configuration)
-
+    
     var configuration: Configuration {
         switch self {
         case .ReadAuthenticatedUser(let config): return config
         }
     }
-
+
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please remove the tab

File: TanukiKit/User.swift [hunk 4]
```diff
@@ -51,28 +90,28 @@ public extension TanukiKit {
...
         case .ReadAuthenticatedUser(let config): return config
         }
     }
-
+    
     var method: HTTPMethod {
         return .GET
     }
-
+
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please remove the tab

File: TanukiKit/User.swift [hunk 5]
```diff
@@ -51,28 +90,28 @@ public extension TanukiKit {
...
     var method: HTTPMethod {
         return .GET
     }
-
+    
     var encoding: HTTPEncoding {
         return .URL
     }
-
+
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please remove the tab

File: TanukiKit/User.swift [hunk 6]
```diff
@@ -51,28 +90,28 @@ public extension TanukiKit {
...
-
+    
     var path: String {
         switch self {
         case .ReadAuthenticatedUser:
             return "user"
         }
     }
-
+
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please remove the tab

File: TanukiKit/User.swift [hunk 7]
```diff
@@ -27,7 +67,6 @@ import RequestKit
 // MARK: request
 
 public extension TanukiKit {
-
```
Comment (pietbrauer) [inline TanukiKit/User.swift:30]: Please add this empty line again

File: TanukiKit/User.swift [hunk 8]
```diff
@@ -3,20 +3,60 @@ import RequestKit
...
+    public var createdAt: String?
+    public var isAdmin: Bool?
     public var bio: String?
-    public var name: String?
+    public var location: String?
+    public var skype: String?
+    public var linkedin: String?
+    public var twitter: String?
+    public var websiteURL: String?
+    public var lastSignInAt: String?
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Shouldn't this be a `NSDate`?

File: TanukiKit/User.swift [hunk 9]
```diff
@@ -3,20 +3,60 @@ import RequestKit
 
 // MARK: model
 @objc public class User: NSObject {
-    public let id: Int
+    public var name: String?
     public var login: String?
+    public let id: Int
+    public var state: String?
+    public var avatarURL: String?
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please use a `NSURL`

File: TanukiKit/User.swift [hunk 10]
```diff
@@ -3,20 +3,60 @@ import RequestKit
...
+    public var webURL: String?
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please use a `NSURL`

File: TanukiKit/User.swift [hunk 11]
```diff
@@ -3,20 +3,60 @@ import RequestKit
...
+    public var createdAt: String?
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please use a `NSDate`

File: TanukiKit/User.swift [hunk 12]
```diff
@@ -3,20 +3,60 @@ import RequestKit
...
+    public var isAdmin: Bool?
     public var bio: String?
-    public var name: String?
+    public var location: String?
+    public var skype: String?
+    public var linkedin: String?
+    public var twitter: String?
+    public var websiteURL: String?
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please use `NSURL`

File: TanukiKit/User.swift [hunk 13]
```diff
@@ -3,20 +3,60 @@ import RequestKit
...
+    public var lastSignInAt: String?
+    public var confirmedAt: String?
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Please use `NSDate`

**Comment by tiferrei** [comment]
Alright, I did the changes for `NSDate` and `NSURL`, but on the tests now shouldn't it check for the same thing, but as NSURL or NSDate, not String? Or does it not matter? For example, this snippet:

``` swift
XCTAssertEqual(subject.webURL, "https://code.tiferrei.com/u/testMcTestface")
XCTAssertEqual(subject.createdAt, "2016-05-03T15:05:46.391Z")
```

Would it still be correct to compare it to an assumed String value because of the quotes? Or will it understand that I want to compare it to this text, but as NSURL?

**Comment by pietbrauer** [comment]
It needs to be the same type. You can check if you did it right by pressing `CMD+U` in Xcode 😉.

**Comment by tiferrei** [comment]
Exactly, by running the tests locally right? The problem is:
1. I accidentally uninstalled Xcode 7 when I installed Xcode 8
2. When I went to reinstall it, I noticed that I ran out of space, (128GB SSD).

So I will upgrade the storage today, that's why I was sending commits with all minimal change, to test it on the CI server, as I couldn't compile Swift 2.2 with Xcode 8 😂
So yeah, forgive all the previous commits but hopefully that won't be a problem from the next 3 hours!

But anyway, as it basically gets the values form the JSON file, which it receives as a String, it should also compare it as a String?

**Comment by tiferrei** [comment]
@pietbrauer Don't think using `NSDate` or `NSURL` will be possible, I tried using them and the tests fail to get it from the JSON, everything as NSDate or NSURL outputs a nil on the tests, change it to String, everything runs perfectly.

**Comment by tiferrei** [comment]
I committed the changes, including the tests that fail with NSDate/NSURL so that you can comment on the code and take a look at the tests on the CI

**Comment by pietbrauer** [comment]
That does obviously not work. You can always take a look at the other libraries, especially Octokit.swift is much further developed than TanukiKit. Maybe this gives you a hint: https://github.com/nerdishbynature/octokit.swift/blob/master/OctoKit/Repositories.swift#L35

**Comment by tiferrei** [comment]
So that should fix everything, tell me if you'd like me to squash the commits again

**Comment by pietbrauer** [comment]
Would be great if you could rebate it to one commit and keep the original commit message so if makes sense in the history.

**Comment by tiferrei** [comment]
I'm sorry, by original commit message do you mean "User Request Improvements"?

**Comment by pietbrauer** [comment]
Hum, yeah maybe something more describing like "Adds additional fields to User model" but feel free to use anything but "squashed" 😉

**Comment by tiferrei** [comment]
Good idea 🙂

**Comment by tiferrei** [comment]
I'm proud, I was able to not mess up squashing them 😀

File: TanukiKit/User.swift [hunk 14]
```diff
@@ -3,20 +3,67 @@ import RequestKit
...
-            name = json["name"] as? String
+            location = json["location"] as? String
+            skype = json["skype"] as? String
+            linkedin = json["linkedin"] as? String
+            twitter = json["twitter"] as? String
+            if let urlString = json["website_url"] as? String, url = NSURL(string: urlString) {
+                websiteURL = url
+            }
+            lastSignInAt = Time.rfc3339Date(json["last_sign_in_at"] as? String)
+            print(lastSignInAt)
```
Comment (pietbrauer) [inline TanukiKit/User.swift]: Can you remove the print? Don't know when that slipped in, sorry for missing it.

Comment (tiferrei) [inline TanukiKit/User.swift]: Oh silly me, I was using it to confirm that it was formatting the time correctly, and forgot to remove it, sorry about that!

**Comment by pietbrauer** [comment]
Can you squash it again 😁

**Comment by tiferrei** [comment]
Here it is 😄

**Comment by pietbrauer** [comment]
This has been a long time coming. 👍

**Comment by tiferrei** [comment]
Well the good thing is that now I already have some kind of "template" for all the other future requests 🙂

**Pull request merged**